### DISCUSSION TITLE
actix-web can't use http2 and http1 simultaneously by listening same address without using https

### DISCUSSION THREAD

**Discussion by rootBiao** [actix/actix-web#2943] [Ideas] [answered]
the lower actix-web is supported,but highter cann't.

**Comment by robjtede** [comment, marked as answer]
Update, support for this was added in actix-http v3.3: https://docs.rs/actix-http/latest/actix_http/struct.HttpService.html#method.tcp_auto_h2c

This is true. Without some form of protocol negotiation like that this is provided by TLS ALPN then we cannot select an appropriate HTTP version to use. Inspecting bytes is possible but would be unnecessary overhead for all Actix Web servers. I do know it is possible to do at the user level but would need to dig up the code to make it work.
